include/databaseFunctions_blog_pic.php

```diff
@@ -1,58 +1,450 @@
-			VALUES('$author', '$title', '$body', '0', '$delKey')
-		")->fetch();
+	}
+
+	function LoginForm(){
+		$errors = array();
+		// var_dump($_REQUEST);
+		if(isset($_REQUEST['login'])){		//NOT PASSING THROUGH IF
+			// var_dump($_REQUEST);
+			$errors+=ValidateTextField('username', $errors);
+			$errors+=ValidateTextField('password', $errors);
+
+			if(UserExists($_REQUEST['username'])){
+				if(GetUser($_REQUEST['username'])['password'] != $_REQUEST['password']){
+					$errors['match'] = 'incorrect password!';
+				}
+			}else{
+				$errors['DNE'] = 'user does not exist!';
+			}
+
+			if(sizeof($errors) == 0){
+				// var_dump($_REQUEST, GetUser($_REQUEST['username'])['userID']);
+				// die();
+				$_SESSION['username'] = $_REQUEST['username'];
+				$_SESSION['userID'] = GetUser($_REQUEST['username'])['userID'];
+				header("Location: /index.php");
+				exit();
+			}else{
+				// var_dump($errors);
+				foreach($errors as $key=>$error){
+					DisplayError($key, $error);
+				}
+				// if(isset($errors['username'])){
+				// 	echo"<div class='required'>Please enter your username.</div>";
+				// }
+				// if(isset($errors['password'])){
+				// 	echo"<div class='required'>Please enter your password.</div>";
+				// }
+				// if(isset($errors['match'])){
+				// 	echo"<div class='required'>Your username and password do not match.</div>";
+				// }
+				// if(isset($errors['DNE'])){
+				// 	echo"<div class='required'>There is no account associated with that username.</div>";
+				// }
+			}
+		}
+		echo"<form method='post'>";
+		ShowTextField(true, 'username');
+		ShowPasswordField('password', 'password');
+		echo"<input type='submit' name = 'login' value='Login'>
+			</form>";
+	}
+
+	function IsLoggedIn(){
+		echo"";
+		if(isset($_SESSION['userID'])){
+			return true;
+		}
+		return false;
+	}
+
+	function PersonalHeading(){
+		echo"
+			<p class='personal'>hi ".$_SESSION['username']."</p>
+		";
+		Logout();
+	}
+
+	function Logout(){
+		if(isset($_REQUEST['logout'])){
+			// echo"<script>
+			// 		alert('Are you sure you want to log out?')
+			// 	</script>
+			// ";
+			session_destroy();
+			header("Location: /index.php");
+			exit();
+		}
+		echo"
+			<form>
+				<input class='personal' type='submit' name='logout' value='Log Out'>
+			</form>
+		";
+	}
+
+	function ValidDelete($postID){
+		// echo"validdelete called";
+		// var_dump($_SESSION);
+		if(isset($_SESSION['userID'])){
+			if($_SESSION['userID'] == GetPostCreator($postID)){
+				return true;
+			}
+			if(GetUser($_SESSION['username'])['userType'] == 'admin'){
+				return true;
+			}
+		}
+		if(GetPostCreator($postID) == NULL){
+			return true;
+		}
+		return false;
+	}
+
+	//TAG DATABASE FUNCTIONS
+	function DeleteTagFromPost($tagname, $postID){		//CHECK DBQUERY SYNTAX
+		$result = dbQuery("
+			DELETE FROM posttags
+			WHERE (SELECT tagname
+				FROM tags
+				INNER JOIN posttags ON
+				tags.tagID = posttags.tagID) = :tagname
```

Comment:
It's hard to be sure without seeing the schema, but it looks to me like you're using the tag name as the primary key instead of a tag id. This has the same issues we discussed earlier with the username.